### PR TITLE
Issue 13 kv username password

### LINKED ISSUE #13: Generate username and password with vault

### PR DESCRIPTION
**Pull request by claesgill**
Closes #13 

- Using Vault to generate username and password for postgres with kv

Followed same structure as in [Minio PR #19](https://github.com/fredrikhgrelland/terraform-nomad-minio/pull/19)

### CODE REVIEW

**Review by pdmthorsrud — COMMENTED** [review]
This is something I forgot to mention in Neha's PR, but I think the playbook setting up the plugin should be called something else than `00-poststart-example.yml`.

**Comment by dangernil** [comment]
> This is something I forgot to mention in Neha's PR, but I think the playbook setting up the plugin should be called something else than `00-poststart-example.yml`.

Maybe something like `00-kv-plugin` ?

**Comment by claesgill** [comment]
> > This is something I forgot to mention in Neha's PR, but I think the playbook setting up the plugin should be called something else than `00-poststart-example.yml`.
> 
> Maybe something like `00-kv-plugin` ?

Prehaps even more spesific: `00-install-vault-secrets-gen-plugin`?

File: 2020-09-16-15-03-53.097-VBoxSVC-21777.log [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Log created: 2020-09-16T15:03:53.974082000Z
+Process ID:  21777 (0x5511)
+Parent PID:  22893 (0x596d)
+Executable:  /usr/lib/virtualbox/VBoxSVC
+Arg[0]: /usr/lib/virtualbox/VBoxSVC
+Arg[1]: --auto-shutdown
+AddRef: illegal refcnt=3221225469 state=2
```
Comment (claesgill) [inline 2020-09-16-15-03-53.097-VBoxSVC-21777.log]: We should probably gitignore this file and delete it from this PR.

File: dev/ansible/02_run_terraform.yml [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+      dest: /tmp
+      mode: 0755
+
+- name: Unarchive a file that is already on the remote machine
+  become: yes
+  unarchive:
+      src: /tmp/terraform_0.13.3_linux_amd64.zip
+      dest: /usr/local/bin
+      remote_src: yes
+
```
Comment (claesgill) [inline dev/ansible/02_run_terraform.yml]: Oh, I forgot to remove this from the PR. I don't think we need this anymore? @dangernil

Comment (dangernil) [reply dev/ansible/02_run_terraform.yml]: You're right

File: dev/ansible/01_generate_secrets_vault.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+- name: Generate Secrets with Vault
+  shell: |
+      vault kv put secret/postgres \
+      username=$(vault write -field='value' gen/password length=12 symbols=0) \
+      password=$(vault write -field='value' gen/password length=12 symbols=0)
+  run_once: true
+  environment:
+      VAULT_ADDR: "http://127.0.0.1:8200"
```
Comment (pdmthorsrud) [inline dev/ansible/01_generate_secrets_vault.yml:8]: Do we actually need this? `VAULT_ADDR` is already set on the machine 
image

File: dev/ansible/01_generate_secrets_vault.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+- name: Delete /etc/default/terraform.tfvars if it exists
+  file:
+      path: /etc/default/terraform.tfvars
+      state: absent
+
+- name: Add username and password to the /etc/default/terraform.tfvars file
+  shell: |
+      echo username='"{{ username.stdout }}"' >> /etc/default/terraform.tfvars
+      echo password='"{{ password.stdout }}"' >> /etc/default/terraform.tfvars
+  run_once: true
```
Comment (pdmthorsrud) [inline dev/ansible/01_generate_secrets_vault.yml]: If we do this from terraform we do not need to create any extra files.

I realise everything I comment ATM was also present in Neha's PR, sorry about that. I glossed over it too quickly it seems!

**Review by pdmthorsrud — DISMISSED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## [0.0.2 UNRELEASED]
 
+### Added
+
+- Using Vault to generate username and password for postgres with kv
```
Comment (zhenik) [inline CHANGELOG.md]:
```suggestion
- Using Vault to generate username and password for postgres with kv #13
```

File: dev/ansible/01_generate_secrets_vault.yml (hunk 3 above)
Comment (zhenik) [reply dev/ansible/01_generate_secrets_vault.yml:8]: I would suggest to leave it explicit. Also, we need to check that `env` variables of user vagrant also available for user ansible.

File: .gitignore [hunk 6]
```diff
@@ -5,4 +5,5 @@ terraform.tfstate*
 .terraform
 .minio.sys
 tmp
-docker/conf/certificates/*.crt
\ No newline at end of file
+docker/conf/certificates/*.crt
+2020*
```
Comment (zhenik) [inline .gitignore]: What is `2020`?

**Review by zhenik — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 6 above)
Comment (dangernil) [reply .gitignore]: It was supposed to handle the timestamp log that gets generated, but maybe it isn't required?

File: dev/ansible/01_generate_secrets_vault.yml (hunk 3 above)
Comment (pdmthorsrud) [reply dev/ansible/01_generate_secrets_vault.yml:8]: Hm, okay, fair enough!

File: .gitignore (hunk 6 above)
Comment (claesgill) [reply .gitignore]: Maybe use `*.log` instead?

File: CHANGELOG.md (hunk 5 above)
Comment (claesgill) [reply CHANGELOG.md]: My bad :facepalm: Do you mind committing this one, @dangernil :smile:

File: .gitignore (hunk 6 above)
Comment (dangernil) [reply .gitignore]: Nevermind, I removed it as the log file isn't there anymore

Comment (pdmthorsrud) [reply .gitignore]: Hm, what made it put up in the first place then? Sounds like it'd be bound to pop up again?

File: dev/ansible/01_generate_secrets_vault.yml (hunk 4 above)
Comment (claesgill) [reply dev/ansible/01_generate_secrets_vault.yml]: How do we do it with terraform?

Comment (pdmthorsrud) [reply dev/ansible/01_generate_secrets_vault.yml]: You can use a `vault_generic_secret`. Hold on, I'll edit this comment with an example (there is one in the code). https://registry.terraform.io/providers/hashicorp/vault/latest/docs/data-sources/generic_secret

EDIT: [here you go](https://github.com/fredrikhgrelland/vagrant-hashistack-template/blob/1a2cc80fadd528596e1024c39e345c05de0ecba5/example/vagrant_box_example/nomad_acl.tf)

**Review by zhenik — APPROVED** [review]

**Comment by Neha-Sinha2305** [comment]
What about updating the documentation?

**Comment by claesgill** [comment]
> What about updating the documentation?

We can do that 👍 

EDIT: 
I think maybe we should prioritize the tasks for this week? Anyways I will create another issue regarding this :)  👉 #16

**Comment by dangernil** [comment]
Can you have another look at this, changed some things regarding secret generation @zhenik @pdmthorsrud

File: dev/.env [hunk 7]
```diff
@@ -1 +1,2 @@
-# User provided environment variables
\ No newline at end of file
+# User provided environment variables
+nomad_acl = true
```
Comment (pdmthorsrud) [inline dev/.env]:
```suggestion
```

File: dev/ansible/01_generate_secrets_vault.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+- name: Register username key
+  shell:
+      cmd: vault kv get -field='username' secret/postgres
+  register: username
+
+- name: Register password key
+  shell:
+      cmd: vault kv get -field='password' secret/postgres
+  register: password
```
Comment (pdmthorsrud) [inline dev/ansible/01_generate_secrets_vault.yml]: You don't need this for anything, do you?
```suggestion
```

File: dev/vagrant/conf/post_bootstrap/00-poststart-example.yml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+- name: set variables
```
Comment (pdmthorsrud) [inline dev/vagrant/conf/post_bootstrap/00-poststart-example.yml]: Change the name of this file, like we discussed earlier

**Review by pdmthorsrud — CHANGES_REQUESTED** [review]

File: dev/vagrant/conf/post_bootstrap/00-poststart-example.yml (hunk 9 above)
Comment (claesgill) [reply dev/vagrant/conf/post_bootstrap/00-poststart-example.yml]: Will do!

File: dev/ansible/01_generate_secrets_vault.yml (hunk 8 above)
Comment (claesgill) [reply dev/ansible/01_generate_secrets_vault.yml]: No, that's right!

File: dev/.env (hunk 7 above)
Comment (claesgill) [reply dev/.env]: 🤦

File: README.md [hunk 10]
```diff
@@ -3,3 +3,16 @@
 ```text
 make test
 ```
+
+## Vault secrets
+The postegres username and password is generated an put in `/secret/postgres` inside Vault.
+
+To get the username and password from Vault you can login to the [Vault-UI](http://localhost:8200/) with token `master` and reveal the username and password in `/secret/postgres`. 
+Alternatively you can you can ssh into the vagrant box with `vagrant ssh`, and use the vault binary to get the username and password. See the following commands:
```
Comment (zhenik) [inline README.md]: typo
```suggestion
Alternatively, you can ssh into the vagrant box with `vagrant ssh`, and use the vault binary to get the username and password. See the following commands:
```

File: example/main.tf [hunk 11]
```diff
@@ -9,8 +13,30 @@ module "postgres" {
...
+  description = "Postgres username"
+  value       = data.vault_generic_secret.postgres_secrets.data.username
+  sensitive   = true
+}
+
+output "postgres_password" {
+  description = "Postgres password"
+  value       = data.vault_generic_secret.postgres_secrets.data.password
+  sensitive   = true
+}
```
Comment (zhenik) [inline example/main.tf]: Why did outputs register in two places? `example/main.tf` and `outputs.tf`.
Do several output variables in `example/main.tf` exist for debugging purposes?

**Review by zhenik — CHANGES_REQUESTED** [review]

File: example/main.tf (hunk 11 above)
Comment (claesgill) [reply example/main.tf]: You are right, we need to remove these. Not really sure why we added them in the first place, but probably debugging purposes.

**Review by zhenik — APPROVED** [review]

File: README.md [hunk 12]
```diff
@@ -3,3 +3,16 @@
 ```text
 make test
 ```
+
+## Vault secrets
+The postegres username and password is generated an put in `/secret/postgres` inside Vault.
```
Comment (pdmthorsrud) [inline README.md]:
```suggestion
The postegres username and password is generated and put in `/secret/postgres` inside Vault.
```

**Review by pdmthorsrud — APPROVED** [review]

**Pull request merged**